packages/astro/src/core/build/plugins/plugin-ssr.ts

```diff
@@ -73,12 +69,224 @@ function vitePluginSSR(
+				if (module) {
+					// we need to use the non-resolved ID in order to resolve correctly the virtual module
+					imports.push(`import * as pageModule from "${virtualModuleName}";`);
+				}
+
+				const ssrCode = generateSSRCode(options.settings.config, adapter);
+				imports.push(...ssrCode.imports);
+				contents.push(...ssrCode.contents);
+
+				return `${imports.join('\n')}${contents.join('\n')}${exports.join('\n')}`;
+			}
+			return void 0;
+		},
+		async generateBundle(_opts, bundle) {
+			// Add assets from this SSR chunk as well.
+			for (const [_chunkName, chunk] of Object.entries(bundle)) {
+				if (chunk.type === 'asset') {
+					internals.staticFiles.add(chunk.fileName);
+				}
+			}
+
+			for (const [chunkName, chunk] of Object.entries(bundle)) {
+				if (chunk.type === 'asset') {
+					continue;
+				}
+				let shouldDeleteBundle = false;
+				for (const moduleKey of Object.keys(chunk.modules)) {
+					if (moduleKey.startsWith(RESOLVED_SERVERLESS_MODULE_ID)) {
+						internals.ssrServerlessEntryChunks.set(moduleKey, chunk);
+						shouldDeleteBundle = true;
+					}
+				}
+				if (shouldDeleteBundle) {
+					delete bundle[chunkName];
+				}
+				// if (chunk.modules[RESOLVED_SSR_VIRTUAL_MODULE_ID]) {
+				// }
+			}
+		},
+	};
+}
+
+export function pluginSSRServerless(
+	options: StaticBuildOptions,
+	internals: BuildInternals
+): AstroBuildPlugin {
+	const ssr =
+		options.settings.config.output === 'server' || isHybridOutput(options.settings.config);
+	return {
+		build: 'ssr',
+		hooks: {
+			'build:before': () => {
+				let vitePlugin =
+					ssr && options.settings.config.build.mode === 'serverless'
+						? vitePluginSSRServerless(internals, options.settings.adapter!, options)
+						: undefined;
+
+				return {
+					enforce: 'after-user-plugins',
+					vitePlugin,
+				};
+			},
+			'build:post': async ({ mutate }) => {
+				if (!ssr) {
+					return;
+				}
+				if (options.settings.config.build?.mode === 'server') {
+					return;
+				}
+
+				if (internals.ssrServerlessEntryChunks.size === 0) {
+					throw new Error(`Did not generate an entry chunk for SSR serverless`);
+				}
+				// TODO: check this commented code
+				// Mutate the filename
+				// internals.ssrEntryChunk.fileName = options.settings.config.build.serverEntry;
+
+				for (const [moduleName, chunk] of internals.ssrServerlessEntryChunks) {
+					const code = await injectManifest(options, internals, chunk);
+					mutate(chunk, 'server', code);
+				}
+			},
+		},
+	};
+}
+
+function generateSSRCode(config: AstroConfig, adapter: AstroAdapter) {
+	const imports: string[] = [];
+	const contents: string[] = [];
+	let middleware;
+	if (config.experimental?.middleware === true) {
+		imports.push(`import * as _middleware from "${MIDDLEWARE_MODULE_ID}";`);
+		middleware = 'middleware: _middleware';
+	}
+	let pageMap;
+	if (config.build.mode === 'serverless') {
+		pageMap = 'pageModule';
+	} else {
+		pageMap = 'pageMap';
+	}
```

Comment:
The diffing doesn't help, but this is the only dynamic part of the code generated in SSR. `pageMap` for server and `pageModule` for serverless